### ISSUE TITLE
`waitCatch` not working as expected.

### ISSUE THREAD

**Issue by gdeest** [barrucadu/dejafu#110]
Hi,

The following code does not work as I expect it:

```haskell
{-# LANGUAGE ScopedTypeVariables #-}
module Main where

import Test.DejaFu (autocheckIO)
import Test.DejaFu.Conc (ConcIO)
import Control.Concurrent.Classy.Async (Async, async, waitCatch)
import Control.Monad.IO.Class (liftIO)

example :: ConcIO ()
example = do
  (a :: Async ConcIO Int) <- async (fail "Error passed through !")
  waitCatch a >>= \e -> liftIO $ case e of
    Left _ -> print "Caught exception."
    Right _ -> print "The impossible happened !"


main :: IO ()
main = do
  autocheckIO example
  return ()
```

I would expect the `IOException` raised by fail to be caught by `waitCatch` ; however, it bubbles up through `autocheckIO` and crashes the program.

Same happens if I replace `fail` with `error` or any other way to throw an exception. Am I doing something wrong, here ?

**Comment by barrucadu** [comment]
Looks like I'm missing an implementation of `fail` for `ConcIO`:

```diff
diff --git a/dejafu/Test/DejaFu/Conc/Internal/Common.hs b/dejafu/Test/DejaFu/Conc/Internal/Common.hs
index 485ad9a..fd67e81 100755
--- a/dejafu/Test/DejaFu/Conc/Internal/Common.hs
+++ b/dejafu/Test/DejaFu/Conc/Internal/Common.hs
@@ -39,9 +39,14 @@ instance Applicative (M n r) where
     pure x  = M $ \c -> AReturn $ c x
     f <*> v = M $ \c -> runM f (\g -> runM v (c . g))

+newtype StringException = StringException String
+  deriving Show
+instance Exception StringException
+
 instance Monad (M n r) where
     return  = pure
     m >>= k = M $ \c -> runM m (\x -> runM (k x) c)
+    fail e = cont (\_ -> AThrow (StringException e))

 -- | The concurrent variable type used with the 'Conc' monad. One
 -- notable difference between these and 'MVar's is that 'MVar's are
```

I'll check what's happening with the `MonadFail` proposal tonight and add the missing definition.

---

`error` definitely won't work, however.

**Comment by barrucadu** [comment]
Fixed in [0.7.1.2](http://hackage.haskell.org/package/dejafu-0.7.1.2/changelog), thanks for reporting.
